pkg/tagset/tags_cache.go

```diff
@@ -0,0 +1,106 @@
+package tagset
+
+// tagsCache caches Tags instances using purpose-specific cache keys.
+//
+// Note that tagsCache instances are not threadsafe
+type tagsCache struct {
+	// number of inserts between rotations
+	insertsPerRotation int
+
+	// number of inserts remaining before next rotation
+	untilRotate int
+
+	// tagests contains the constitutent tagset maps. This is a slice of length
+	// cacheCount.  The first map is the newest, into which new values will be
+	// inserted.
+	maps []map[uint64]*Tags
+}
+
+func newTagsCache(insertsPerRotation, cacheCount int) tagsCache {
+	maps := make([]map[uint64]*Tags, cacheCount)
+	for i := range maps {
+		maps[i] = make(map[uint64]*Tags)
+	}
+	return tagsCache{
+		insertsPerRotation: insertsPerRotation,
+		untilRotate:        insertsPerRotation,
+		maps:               maps,
+	}
+}
+
+// getCachedTags gets an element from the cache, calling miss() to generate the
+// element if not found.
+func (tc *tagsCache) getCachedTags(key uint64, miss func() *Tags) *Tags {
+	v, ok := tc.search(key)
+	if !ok {
+		v = miss()
+		tc.insert(key, v)
+	}
+	return v
+}
+
+// getCachedTagsErr is like getCachedTags, but works for miss() functions that can
+// return an error.  Errors are not cached.
+func (tc *tagsCache) getCachedTagsErr(key uint64, miss func() (*Tags, error)) (*Tags, error) {
+	v, ok := tc.search(key)
+	if !ok {
+		var err error
+		v, err = miss()
+		if err != nil {
+			return nil, err
+		}
+		tc.insert(key, v)
+	}
+	return v, nil
+}
+
+// search searches for a key in maps older than the first.  If found, the key
+// is copied to the first map and returned.
+func (tc *tagsCache) search(key uint64) (*Tags, bool) {
+
+	v, ok := tc.maps[0][key]
+	if ok {
+		return v, true
+	}
+
+	cacheCount := len(tc.maps)
+	for i := 1; i < cacheCount; i++ {
```

Comment:
It will depend on the mix of tagset lifetimes, which may vary from usage to usage.  I suspect in most cases cacheCount = 2 will be the right choice, but if there's a mix of 50% one-time tagsets and 50% long-lived tagsets, then maps[1] gives a chance to "rescue" those long-lived tagsets and pull them back into maps[0].  cacheCount = 1 would be like the current string interner, just periodically deleting the entire cache -- not very efficient.

I expect this is something we'll want to tune, at first by hand and probably later automatically.  @vickenty had some measurements of this "churn rate" for one prod cluster, that can provide a starting point.